File: APIs/Search/SearchAPIs/AzureCognitiveServices/S2Search.Search.API/Domain/Models/Request/SearchRequest.cs
﻿using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace Domain.Models.Request
{
    public class SearchRequest
    {
        [Required]
        [JsonPropertyName("searchTerm")]
        public string SearchTerm { get; set; }

        [Required]
        [JsonPropertyName("filters")]
        public string Filters { get; set; }

        [Required]
        [JsonPropertyName("orderBy")]
        public string OrderBy { get; set; }

        [Required]
        [JsonPropertyName("pageNumber")]
        public int PageNumber { get; set; }

        [Required]
        [JsonPropertyName("pageSize")]
        public int PageSize { get; set; }

        [JsonPropertyName("numberOfExistingResults")]
        public int NumberOfExistingResults { get; set; }

        [JsonPropertyName("callingHost")]
        public string CallingHost { get; set; }
    }
}

File: UIs/SearchUIs/AzureCognitiveServices/S2Search.Search.NextJS.ReactUI/S2Search/pages/api/shared/request/SearchRequest.js
import {
  DefaultPageNumber,
  DefaultPageSize,
} from '../../../../common/Constants';

class SearchRequest {
  constructor(
    searchTerm,
    filters,
    orderBy,
    pageNumber,
    pageSize,
    numberOfExistingResults,
    callingHost
  ) {
    if (searchTerm === undefined || searchTerm === null || searchTerm === ' ') {
      this.searchTerm = '';
    } else {
      this.searchTerm = searchTerm;
    }

    if (filters === undefined || filters === null) {
      this.filters = [];
    } else {
      this.filters = filters;
    }

    if (orderBy === undefined || orderBy === null) {
      this.orderBy = ['price desc'];
    } else {
      this.orderBy = orderBy;
    }

    if (pageNumber === undefined || pageNumber === null) {
      this.pageNumber = DefaultPageNumber;
    } else {
      this.pageNumber = pageNumber;
    }

    if (pageSize === undefined || pageSize === null) {
      this.pageSize = DefaultPageSize;
    } else {
      this.pageSize = pageSize;
    }

    if (
      numberOfExistingResults === undefined ||
      numberOfExistingResults === null ||
      filters.length == 1
    ) {
      this.numberOfExistingResults = pageSize;
    } else {
      this.numberOfExistingResults = numberOfExistingResults;
    }

    if (callingHost === undefined || callingHost === null) {
      this.callingHost = [];
    } else {
      this.callingHost = callingHost;
    }
  }

  // Convert SearchRequest instance to a plain object for Redux serialization
  toPlainObject() {
    return {
      searchTerm: this.searchTerm,
      filters: this.filters,
      orderBy: this.orderBy,
      pageNumber: this.pageNumber,
      pageSize: this.pageSize,
      numberOfExistingResults: this.numberOfExistingResults,
      callingHost: this.callingHost,
    };
  }

  // Create SearchRequest instance from plain object
  static fromPlainObject(obj) {
    if (!obj) return null;
    return new SearchRequest(
      obj.searchTerm,
      obj.filters,
      obj.orderBy,
      obj.pageNumber,
      obj.pageSize,
      obj.numberOfExistingResults,
      obj.callingHost
    );
  }
}

export default SearchRequest;
